### DISCUSSION TITLE
Other distros ?

### DISCUSSION THREAD

**Discussion by laorka** [1000001101000/Debian_on_Buffalo#101] [General]
In your opinion, if it works well with Debian, would it work with others distros like Cent OS ?

**Comment by 1000001101000** [comment]
Good to hear from you!

It depends in part on which device you’re interested in.

It looks like they have an armv7 port which could theoretically work with the armada 370/xp devices. 

You’d need to determine if their kernel has support for those SoCs compiled in and if not compile a custom one.

You would also have to make a suitable device-tree for the kernel the OS uses. It would probably work with the closest Debian version’s tree but may need some tweaking.

You would also need an initrd which handles/ignores the kernel cmdline set by the stock bootloader and a way to package initrd and kernel images for uboot. Debian’s flash-kernel tool handles all that seamlessly, I don’t know what options exist for other distros.

Once you solved the issues required to make it boot and handle things like kernel updates I would imagine it would run similarly to how it does on Debian.

**Comment by laorka** [comment]
So it seems like a good project to learn about boot, devices-tree and init stuff (I haven't looked at those yet)

I'll try it on a LS220DR (SoC supported by CentOS), if it works I'll put it on GitHub

**Comment by 1000001101000** [comment]
I'd say it depends on your current skillset and learning style. 

Doing the same on the ls421de on Debian was pretty much exactly how I learned a lot of this stuff, and understanding these pieces of the boot process has helped me a lot on PCs and other systems. 

I managed to get these devices working without a serial console on that device but that was starting with a rootfs that already worked and with a lot of trial and error getting the initrd working. It might be a bit harder to get started in this case without some of those elements in place. Though you might be able to do it in steps such as getting your debian rootfs to boot using the centos kernel first or something like that.

You could also try setting up the serial console which will make things much easier for the boot part. I haven't done so on this model yet but I have some details about it form someone who did:
https://buffalonas.miraheze.org/wiki/Linkstation_LS220D#Serial_Console

**Comment by RQWorldblender** [comment]
Hi, I've come here as I thought this would be the most relevant place to discuss what I would like to do - **I have a Buffalo Linkstation LS-410D running Debian 10 Buster, courtesy of your installers. I have been thinking of replacing this install with Ubuntu 20.04 LTS or 22.04 LTS instead, as I already use this distro on my main desktop and laptop computers, and it would help create more consistency in managing these systems.** For example, I may be able to take advantage of Canonical's services that allow me to access security updates for Ubuntu LTS releases even after their EOL, and that can't be done on upstream Debian. This distro has ports for both 32-bit and 64-bit ARM, and recent releases have official images for the Raspberry Pi, and this device has 32-bit ARM.

Older releases did have a legacy installer similar to the Debian one, here's the one for 20.04 LTS (newer LTS releases don't ship this image anymore): http://ports.ubuntu.com/dists/focal/main/installer-armhf/current/legacy-images/

Otherwise, the 22.04 LTS release has preinstalled images meant for Raspberry Pi boards, but which could work on other ARM hardware: http://cdimage.ubuntu.com/releases/jammy/release/

I have already checked to see if the flash-kernel package exists in other distros at https://pkgs.org/download/flash-kernel, and it exists in both Debian and Ubuntu, but not yet in non-DEB-based (especially with RPM-based distros) distros. Now as for actually converting or replacing the Debian install to/with Ubuntu, I could do this whole process by hand, but I would risk system breakage in case something goes wrong. I could try to create a bootable image all by hand, but that can quite time-consuming especially if my end result doesn't work. The hard part here is creating a suitable uBoot image that will actually boot, and it seems that it isn't easy for me, because Ubuntu is likely not targeted at any these Buffalo NAS devices I'm trying to run it on (but I'd like to try anyways).

**What's the best and/or easiest way I can get Ubuntu (being a Debian derivative) running on my Buffalo Linkstation 410D (and similar devices), even if I have to do a little tinkering (but not so much that it can cause frustration)?**

**Comment by 1000001101000** [comment]
Good to hear from you!

It should be possible to use my debootstrap script by adjusting the url and distro information:
https://github.com/1000001101000/Debian_on_Buffalo/wiki/Alternate-install-method-via-debootstrap-script

There may be some other tweaks required.

**Reply by RQWorldblender** [reply]
Here, I modified that script so that it will fetch from the Ubuntu repositories instead, but I haven't tried it out yet. I'll test it soon, but you can do so too if you want. Remove the .txt extension to make it work as a shell script.

[generate_disk_image_ubuntu.sh.txt](https://github.com/1000001101000/Debian_on_Buffalo/files/9843301/generate_disk_image_ubuntu.sh.txt)

**Comment by 1000001101000** [comment]
I gave it a try and was eventually able to adjust the script to generate a working image for my TS3400D ([generate_disk_image.sh.txt](https://github.com/1000001101000/Debian_on_Buffalo/files/9853663/generate_disk_image.sh.txt)) though with a few issues:

- Finding an Ubuntu mirror with armhf packages was difficult, though I found one with a variation of this script: https://gist.github.com/mtnygard/9cfef214c6942171a578da11741a9d85
- I had to install Ubuntu's version of debootstrap but that worked without issue on my Debian PC.
- Had to bump up the rootfs size to accommodate the larger Ubuntu install
- Ubuntu doesn't appear to support ifup/ifdown style network config by default, for this test I configured using systemd-networkd since that was included but it might not be the best way.
- My "fix" for the onboard MAC address depends on an ifup script so currently the device will get a random MAC address at boot. 
- Ubuntu names their kernel `linux-image-generic` rather than `linux-image-armmp` which requires a minor tweak, however....
- **I couldn't get Ubuntu's kernel to boot.** I didn't get any console output it just seemed to hang. It's not obvious to me what the issue is though it does have a fairly different configuration.
- I added some code to have it download and install Debian's kernel for testing purposes which does result in a working image.



```
debian@192.168.1.177's password: 
Welcome to Ubuntu 22.04 LTS (GNU/Linux 5.10.0-18-armmp armv7l)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage

The programs included with the Ubuntu system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Ubuntu comes with ABSOLUTELY NO WARRANTY, to the extent permitted by
applicable law.

debian@ts3400d:~$ cat /etc/os-release 
PRETTY_NAME="Ubuntu 22.04 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04 (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```

**Reply by RQWorldblender** [reply]
Sorry that I didn't get to test my changes until you made yours, but from what you did, I can tell that my initial effort wouldn't have been enough to generate a working image. For the Ubuntu mirror with armhf or arm64 images, I suggest using http://ports.ubuntu.com/ as the primary mirror. About setting the MAC address by ifup script, maybe it could be possible to convert it into a format that can be used by systemd?

Let me know if there's something else I should do so that we can provide Ubuntu as an alternative to installing Debian.

Okay, I did find an answer on how to set a permanent MAC address using systemd, going through netplan: https://askubuntu.com/a/1395047/1052069
The summary is:
1. Create the file `/etc/netplan/config.yaml`
2. Its contents should look like for example (only one or two ethernet lines are needed):
```
network:
    version: 2
    renderer: networkd
    ethernets:
        eth0:
            macaddress: ba:e4:64:bc:81:d7
        eth1:
            macaddress: c2:1c:fb:6b:fb:73
        eth2:
            macaddress: 2A:77:19:7C:B4:E4
        eth3:
            macaddress: ba:2a:5c:54:35:60
```
3. Apply settings with `sudo netplan apply` (only if editing this configuration on-device)

**Comment by 1000001101000** [comment]
I know from experience it's a lot easier to get started on this type of project if you have something that at least boots successfully to start with. I figured I would see if I could get you at least that far, even though it ended up being more difficult than I expected.  

I think your first step should be to adjust my script to your device model and try generating an image and see if it will boot successfully. From there you can poke around at the Ubuntu install and see if it satisfies your needs.

For Network and MAC addresses:

- You may want to adjust the debootstrap script to generate a netplan yaml rather than the systemd-networkd unit file I went with. I did it that way only because I was familiar with that method (and it was pretty simple) you may want to do it differently to conform to Ubuntu defaults etc.
- My script for setting the MAC is here: https://github.com/1000001101000/Debian_on_Buffalo/blob/master/Tools/ifup-mac.sh . It is a mess partly because it's an ugly hack and partly because I had to keep making it uglier to maintain compatibility with some driver changes etc. You should be able to set something up with either netplan or systemd-networkd to call it after the interfaces come up. 
- There is a kernel patch out there that deals with this though I think it's unsuitable to get included into mainline because it's specific to a uboot/device-tree combination....or something https://github.com/tohenk/linkstation-mod/blob/master/lskernel/patches/4.19/0002-ARM-atags-fdt-retrieve-MAC-addresses-from-Marvell-bo.patch
- I have a different fix I'm getting ready to deploy which side-steps the need for any of this by adding the MAC(s) to the device-tree early in the boot process, though that introduces it's own challenges.  

Ubuntu Kernel:

- I expect that the issue can be corrected by adjusting the configuration and re-compiling. 
- There is a slight chance that Ubuntu applies some patch to their kernel which breaks compatibility with my device-trees or otherwise breaks support for these devices, but I doubt it. 
-  Usually you can find the closest working kernel to the one you want and compare the configs to find what the issue is, often with some trail-and-error testing to identify what the problem is.
- Once you identify the issue with the Ubuntu kernel you might be able to get them to adjust their kernel config (if it's just some missing option). Failing that you can always just maintain your own kernel.

**Reply by RQWorldblender** [reply]
Sorry for a late response. but I now tried your modified script. It worked for the most part, but there are a few but significant errors I got during the install process:
```
cp: cannot stat '../../Jammy/device_trees/*.dtb': No such file or directory
device-tree install failed
[...]
downloading Debian Bullseye kernel since Ubuntus didnt seem to work
tail: invalid number of lines: ‘+’
basename: missing operand
Try 'basename --help' for more information.
dpkg: error: --install needs at least one package archive file argument

Type dpkg --help for help about installing and deinstalling packages [*];
Use 'apt' or 'aptitude' for user-friendly package management;
Type dpkg -Dhelp for a list of dpkg debug flag values;
Type dpkg --force-help for a list of forcing options;
Type dpkg-deb --help for help about manipulating *.deb files;

Options marked [*] produce a lot of output - pipe it through 'less' or 'more' !
```
There has to be some way around these errors, but my existing Debian Buster setup is running fine, and it may also be likely that I replace my Buffalo Linkstation LS-410D with something higher-end or more recent (perhaps something with ARMv8), as this device's CPU lacks NEON support (that certain programs like Firefox and NodeJS require in their stock builds for ARM architectures). I think that this could be one of the causes as to why the stock Linux image shipped with Ubuntu didn't fully boot. I think I have been yet again effectively trapped just because of a missing CPU feature, that being NEON.

**Comment by 1000001101000** [comment]
The device-tree message is just because there aren't dedicated "Jammy" device trees in my repo, the script will just use the current Debian Stable in that event. 

The other errors are because my logic for finding and downloading the current Debian kernel failed for some reason. I'll take a look at that and see if I get the same result here. 

I had some luck using some GUI apps on an LS210 a while back over X11 forwarding and/or VNC, but I would think FireFox would be a pretty bad match for this device just based on memory requirements

**Comment by 1000001101000** [comment]
[generate_disk_image.sh.txt](https://github.com/1000001101000/Debian_on_Buffalo/files/9953266/generate_disk_image.sh.txt)
The Debian kernel install issue was an easy fix (see attached)
